### ISSUE TITLE
Support substituting environment variables in docker-compose file

### ISSUE THREAD

**Issue by amsb** [convox/rack#1607]
Docker Compose supports substituting environment variables in Compose files:

* [Substituting environment variables in compose files](https://docs.docker.com/compose/environment-variables/#/substituting-environment-variables-in-compose-files)
* [Variable substitution](https://docs.docker.com/compose/compose-file/#variable-substitution)

Supporting this in Convox would provide a convenient tactic for maintaining parity between various environments. For example, the following `docker-compose.yml` file:

```yaml
version: '2'

services:
  web:
    build: .
    command: serve.py
    ports:
      - 80:80
    links:
      - postgres
    environment:
      - POSTGRES_URL="postgres://${PGUSER}:${PGPASSWORD}/${PGDATABASE}"

  postgres:
    image: convox/postgres
    ports:
      - 5432
    environment:
      - POSTGRES_USER=${PGUSER}
      - POSTGRES_PASSWORD=${PGPASSWORD}
      - POSTGRES_DATABASE=${PGDATABASE}
      - LINK_USERNAME=${PGUSER}
      - LINK_PASSWORD=${PGPASSWORD}
      - LINK_PATH="/${PGDATABASE}"
```

along with this `.env` file:

```
PGUSER=TESTUSER
PGPASSWORD=TESTPASS
PGDATABASE=myapp
```

would have the following advantages:

1. Set postgres user, password, and database only once, in `.env`.
2. Better control accidentally leaking insecure credentials to production.
3. Use same files for both Docker Compose and Convox.

In this particular example, there is the disadvantage that deployment would require explicitly setting environment values for a service that would be replaced by a resource and scaled to 0.

The 2.1 Docker Compose file format also supports `${VARIABLE:-default}` and `${VARIABLE-default}` inline default value syntax.

**Comment by mwarkentin** [comment]
@amsb Generally we use the built-in [`convox env`](https://convox.com/docs/environment/) commands to manage configuration like your example - do you see some benefits to doing it this way instead?

**Comment by amsb** [comment]
@mwarkentin The primary advantages I see emanate from maintaining the local development environment where the `convox env` doesn't apply. Here's a specific example I shared on Slack. The `convox/postgres` image sets the LINK vars to the root postgres user while my applications do not connect to the database as the root user (even in development). This leads me into a bit of a naming kerfuffle in that I can either override the LINK vars in `docker-compose.yml` where it is a bit ugly to have passwords (even if for development) hard-coded and checked-in (as well as momentarily in production when deployed but before the service is scaled down), I can set `LINK_USERNAME` in `.env` where it is a bit unclear and may conflict with other linked services that might require the same treatment or I can ignore the automatically generated `POSTGRES_URL` environment variable and create it explicitly from (for example) combining `POSTGRES_HOST` LINK var and my own `PGUSER` but then I’m working against some of the conventions.

**Comment by mwarkentin** [comment]
Ah, right - I remember you mentioning the different DB user.

**Comment by amsb** [comment]
I'm also maintaining a `docker-compose.test.yml` file for running tests with `docker-compose run —rm web run_tests.py` and a `docker-compose.yaml` for running with Convox since `convox start` did't return the exit status and the convox environment didn't have enough information (due to the LINK system) for Docker Compose. Those are my two use cases at the moment although I'll admit that it is lower priority than working on other projects like ALB and service auto-scaling since maintaining multiple Docker Compose files isn't that hard.

**Comment by ffjia** [comment]
@dlanger 
Could you elaborate more about how you set the environment variable? I did some tests, trying to set `image' from a variable, all failed :(

**Comment by dlanger** [comment]
Sorry - I no longer have access to that code to consult it, and I don't
remember offhand.
